src/vmware/pvscsi.c

```diff
@@ -0,0 +1,664 @@
+#include <kernel.h>
+#include <page.h>
+#include <pci.h>
+#include "pvscsi.h"
+#include "storage.h"
+#include "virtio/scsi.h"
+
+#ifdef PVSCSI_DEBUG
+# define pvscsi_debug rprintf
+#else
+# define pvscsi_debug(...) do { } while(0)
+#endif // defined(PVSCSI_DEBUG)
+
+#define PVSCSI_DEFAULT_NUM_PAGES_REQ_RING 8
+#define PVSCSI_CDB_SIZE 16
+#define PVSCSI_SENSE_SIZE 256
+
+struct pvscsi_hcb {
+    struct pvscsi_ring_req_desc    *e;
+    struct list links;
+
+    u16 target;
+    u16 lun;
+    u32 data_len;
+    u64 data_addr;
+
+    u8 cdb[PVSCSI_CDB_SIZE];
+    u16 host_status;
+    u16 scsi_status;
+
+    thunk completion;
+    u32 alloc_len;                     // allocated data length
+    void *data;                       // allocated data
+
+    u8 sense[PVSCSI_SENSE_SIZE];
+};
+
+typedef struct pvscsi {
+    struct pci_dev _dev;
+    pci_dev dev;
+
+    struct pci_bar bar;
+
+    heap contiguous;
+    heap general;
+
+    heap hcb_objcache;
+    struct spinlock mem_lock;
+
+    struct pvscsi_rings_state    *rings_state;
+    struct pvscsi_ring_req_desc    *req_ring;
+    struct pvscsi_ring_cmp_desc    *cmp_ring;
+
+    u32 max_targets;
+    u32 adapter_queue_size;
+
+    struct list hcb_queue;
+    struct spinlock queue_lock;
+
+    thunk intr_handler;
+    thunk rx_service;
+    queue rx_servicequeue;
+
+    u16 target;
+    u16 lun;
+    u64 capacity;
+    u64 block_size;
+} *pvscsi;
+
+static void pvscsi_write_cmd(pvscsi dev, u32 cmd, void *data, u32 len)
+{
+    assert(len % sizeof(u32) == 0);
+
+    u32 *data_ptr = data;
+    len /= sizeof(u32);
+
+    pci_bar_write_4(&dev->bar, PVSCSI_REG_OFFSET_COMMAND, cmd);
+    for (int i = 0; i < len; ++i) {
+        pci_bar_write_4(&dev->bar, PVSCSI_REG_OFFSET_COMMAND_DATA, data_ptr[i]);
+    }
+}
+
+static void pvscsi_hcb_dealloc(pvscsi dev, struct pvscsi_hcb *hcb)
+{
+    u64 flags = spin_lock_irq(&dev->mem_lock);
```

Comment:
Does this need to be irq-safe? It seems like the int handler wouldn't touch the hcb cache.